text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,106 @@
+
+In order to solve the above problems, this proposal presents another kv interface - VerKV, which mainly adds version information such as timestamp to the key. At the same time, this interface will be designed to be approximate to the semantics of HBase, which will facilitate the service migration from HBase to TiKV, thereby integrating TiKV better into the HBase ecosystem.
+
+## Detailed designs
+
+### Functions and scenarios
+
+Here are the functionality designs and the corresponding application scenarios for VerKV:
+   - Support data migration, incremental backup, CDC (Change Data Capture) and cross-IDC cluster synchronization application scenarios include:
+     - Online data replication between TiKV clusters, which includes a full migrations followed by incremental migrations;
+     - Cluster data backup, which includes a first full backup followed perform an incremental data backups every other time;
+     - CDC service created based on incremental data acquisition; can reliably provide incremental data services;
+     - In cross data center cluster synchronization scenarios, VerKV can not only process  incremental data synchronization, but also handle write conflicts caused by the same key.
+
+  - Support multi-versioned data. The corresponding application scenarios include: 
+    - Data version retrieval. For certain businesses, multiple versions of historical data must be maintained for for backtracking and analysis. Reliance on third-party storage and consumption of computing resources (user dynamic characteristics, User behavior link tracking) needs to be reduced, and the data processing process optimized;
+    - Data version control, for fast data recovery after problematic data writes in businesses.
+
+  - Support data expiration based on TTL. The corresponding application scenarios include:
+    - Setting expiration time to obtain data of a recent period
+    - Setting expiration time to clear inactive data for a specified amount of time
+
+
+### Compatibility
+
+The design adds a new Column Family called `VerCF` in RocksDB, which avoids poential compatibility issues with the existing RawKV annd TxnKV data and processing logics, and could be created dynamically when needed. This design allows maximum reuse of the current implementations of TiKV and PD.
+
+### Data model and interface
+
+VerKV is a kv-like interface similar to the RawKV interface. It hides version information from upper-layer services. Its data model is multi-versioned KV between RawKV and TxnKV. 
+
+#### Data model
+
+The data model is like `key:version -> value`. The version uses the current timestamp by default, or can be specified by the user (such as the global `ts` obtained from the PD TSO). And the implementation is subject to further change if we use the user TS feature of RocksDB.
```

Comment:
I suggest keeping only one time-source, PD. Different time-sources may require different implementation of the CDC, it could be difficult.

Reply:
for strong consistency, it is good to use pd tso. For weak consistency, is it possible for cdc to handle it flexibly?

Reply:
> For weak consistency, is it possible for cdc to handle it flexibly?

It depends on whether we accept data loss/duplication on CDC service. With weak time-source, it could be difficult or impossible to prevent data loss/duplication.

The behavior of backup and CDC could be undefined if we allow using different time sources.